utils/prof_utils.py

```diff
@@ -0,0 +1,1150 @@
+
+    for j in range(0, len(popt), 3):
+        z = multi_gauss(x, *popt[j:j+3])
+        plt.plot(x, z, "--", label="Gaussian Component {}".format(int((j+3)/3)))
+    if maxima:
+        for i, mx in enumerate(maxima):
+            plt.axvline(x=(mx + maxima_e[i]), ls=":", lw=2, color="gray")
+            plt.axvline(x=(mx - maxima_e[i]), ls=":", lw=2, color="gray")
+
+    plt.title(plot_name.split("/")[-1].split(".")[0], fontsize=22)
+    plt.xticks(fontsize=18)
+    plt.yticks(fontsize=18)
+    plt.xlim(0, len(y))
+    plt.xlabel("Bins", fontsize=20)
+    plt.ylabel("Intensity", fontsize=20)
+
+    plt.plot(x, y, label="Original Profile", color="black")
+    plt.plot(x, fit, label="Gaussian Model", color="red")
+    plt.legend(loc="upper right", prop={'size': 16})
+    plt.savefig(plot_name)
+    plt.close()
+
+#---------------------------------------------------------------
+def fit_gaussian(profile, max_N=6, min_comp_len=0, plot_name=None, alpha=3.):
+    """
+    Fits multiple gaussian components to a pulse profile and finds the best number to use for a fit.
+    Will always fit at least one gaussian per profile component.
+    Profile components are defined by find_components().
+    Each gaussian is defined by the following: y = amp * np.exp( -((x - ctr)/wid)**2)
+
+    Parameters:
+    -----------
+    profile: list
+        A list containing the profile data
+    max_N: int
+        OPTIONAL - The maximum number of gaussain components to attempt to fit. Default: 6
+    min_comp_len: float
+        OPTIONAL - Minimum length of a component to be considered real. Measured in bins. Default: 0
+    plot_name: string
+        OPTIONAL - If not none, will make a plot of the best fit with this name. Default: None
+    alpha: float
+        OPTIONAL - The alpha value to be used in sigmaClip(). Default: 3
+
+    Returns:
+    --------
+    [fit, redchisq, best_bic, popt, pcov]: list
+        fit: list
+            The data containing the multi-component gaussian fit to the input profile
+        redchisq: float
+            The reduced chi-sqaured value of the fit
+        best_bic: float
+            The bayesian information criterion for the fit
+        popt: list
+            A list of floats where each 3 numbers describes a single gaussain and are 'ctr', 'amp' and 'wid' respectively
+        pcov: numpy matrix
+            The covariance matrix generated by the curve_fit function
+    """
+    #chi sqaured evaluation
+    def chsq(observed_values, expected_values, err):
+        test_statistic=0
+        for observed, expected in zip(observed_values, expected_values):
+            test_statistic+=((float(observed)-float(expected))/float(err))**2
+        return test_statistic
+
+    #Take noise mean and normalize the profile
+    _, clipped = sigmaClip(profile, alpha=alpha)
+
+    #Check the clipped profile
+    check_clip(clipped)
+
+    y = np.array(profile) - np.nanmean(np.array(clipped))
+    max_y = max(y)
+    len_y = len(y)
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [local variable 'len_y' is assigned to but never used](https://app.codacy.com/manual/CIRA-Pulsars-and-Transients-Group/vcstools/pullRequest?prid=4645369)